examples/pii_detection_presidio.py

```diff
@@ -0,0 +1,161 @@
+# Databricks notebook source
+# MAGIC %md
+# MAGIC # PII detection with DiscoverX & Presidio
+# MAGIC
+# MAGIC This notebooks uses [DiscoverX](https://github.com/databrickslabs/discoverx) to run PII detection with [Presidio](https://microsoft.github.io/presidio/) over a set of tables in Unity Catalog.
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Install dependencies
+
+# COMMAND ----------
+
+# MAGIC %pip install presidio_analyzer
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Download detection model
+
+# COMMAND ----------
+
+# MAGIC %sh python -m spacy download en_core_web_lg
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Define variables
+
+# COMMAND ----------
+
+# TODO: Change the table selection
+from_tables = "sample_data_discoverx.*.*"
+
+# TODO: Change the num of rows to sample
+sample_size = 100
+
+
+# COMMAND ----------
+
+from presidio_analyzer import AnalyzerEngine, PatternRecognizer
+from pyspark.sql.functions import pandas_udf, col, concat, lit, explode, count, avg, sum
+from pyspark.sql.types import ArrayType, StringType, StructType, FloatType, StructField
+import pandas as pd
+
+# COMMAND ----------
+
+from discoverx import DX
+
+dx = DX()
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Transform all sampled tables to long format
+# MAGIC
+# MAGIC This will take all columns of type string and unpivot (melt) them into a long format dataset
+
+# COMMAND ----------
+
+unpivoted_df = (
+    dx.from_tables(from_tables)
+    .unpivot_string_columns(sample_size=sample_size)
+    .to_union_dataframe()
+    .localCheckpoint()  # Checkpointing to reduce the query plan size
+)
+
+# unpivoted_df.display()
+
+# COMMAND ----------
+
+unpivoted_stats = unpivoted_df.groupBy("table_catalog", "table_schema", "table_name", "column_name").agg(
+    count("string_value").alias("sampled_rows_count")
+)
+
+# unpivoted_stats.display()
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Define Presidio UDFs
+
+
+# COMMAND ----------
+
+# Define the analyzer, and add custom matchers if needed
+analyzer = AnalyzerEngine()
+
+# broadcast the engines to the cluster nodes
+broadcasted_analyzer = sc.broadcast(analyzer)
```

Comment:
I think using the Iterator -> Iterator pandas_udf is even better than broadcasting ...
https://spark.apache.org/docs/3.1.2/api/python/reference/api/pyspark.sql.functions.pandas_udf.html

Reply:
Dome